[10.0-preview.1] ASP.NET  `RouteHandlerAnalyzer`  -  `System.IndexOutOfRangeException`

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

I see some warnings like: 

```
CSC : warning AD0001: Analyzer 'Microsoft.AspNetCore.Analyzers.RouteHandlers.RouteHandlerAnalyzer' threw an exception of type 'System.IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'. [c:\Users\erwin\bank-api\BankApi.Service.Stable\BankApi.Service.Stable.csproj]
```

When i build the .NET 10 version of my project: https://github.com/erwinkramer/bank-api/tree/feature/dotnet10, it didn't happen on .NET 9. See the diff here: https://github.com/erwinkramer/bank-api/pull/6/files

### Expected Behavior

Should just be as .NET 9, no warnings. 

### Steps To Reproduce

Build .NET 9 version on: https://github.com/erwinkramer/bank-api/tree/main - this doesn't throw the warnings
Build .NET 10 version on: https://github.com/erwinkramer/bank-api/tree/feature/dotnet10 - this throws the warnings


### Exceptions (if any)

_No response_

### .NET Version

10.0.100-preview.1.25120.13

### Anything else?

Using VS Code

@captainsafia thanks, I'm willing to give it a try. I started with https://github.com/dotnet/aspnetcore/blob/main/docs/BuildFromSource.md, but at:

```
cd src/Framework
./build.cmd
```

I'm getting:

```
 Microsoft.AspNetCore.App.Runtime failed with 1 error(s) (0,0s)
    C:\Users\erwin\.nuget\packages\microsoft.dotnet.build.tasks.installers\10.0.0-beta.25106.4\build\wix\wix.targets(149,11): error MSB4006: There is a circular dependency in the target dependency graph involving target "PublishFilesForWixInstaller".
```

I assumed i maybe needed the Wix dependency, as specified here as optional, so i installed it anyway, but no effect: https://github.com/dotnet/aspnetcore/blob/main/docs/BuildFromSource.md#optional-dependencies

I'm using `main` @ 1f82fa558dd7afb0ac94a43c7b80f40499bad98c

--- a/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
+++ b/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
@@ -30,7 +30,7 @@ public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
                 continue;
             }
 
-            var parameterTypeSymbol = ResovleParameterTypeSymbol(handlerDelegateParameter);
+            var parameterTypeSymbol = ResolveParameterTypeSymbol(handlerDelegateParameter);
 
             // If this is null it means we aren't working with a named type symbol.
             if (parameterTypeSymbol == null)
@@ -113,7 +113,7 @@ public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
             return false;
         }
 
-        static INamedTypeSymbol? ResovleParameterTypeSymbol(IParameterSymbol parameterSymbol)
+        static INamedTypeSymbol? ResolveParameterTypeSymbol(IParameterSymbol parameterSymbol)
         {
             INamedTypeSymbol? parameterTypeSymbol = null;
 

--- a/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
+++ b/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
@@ -128,7 +128,8 @@ public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
             }
 
             // If it is nullable, unwrap it.
-            if (parameterTypeSymbol!.ConstructedFrom.SpecialType == SpecialType.System_Nullable_T)
+            if (parameterTypeSymbol!.ConstructedFrom.SpecialType == SpecialType.System_Nullable_T &&
+                parameterTypeSymbol.TypeArguments.Length > 0)
             {
                 parameterTypeSymbol = parameterTypeSymbol.TypeArguments[0] as INamedTypeSymbol;
             }

--- a/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
+++ b/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
@@ -127,7 +127,8 @@ public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
                 parameterTypeSymbol = namedTypeSymbol;
             }
 
-            // If it is nullable, unwrap it.
+            // If it is nullable and we have type arguments, unwrap it.
+            // The length check aims to alleviate AD0001 warnings when referencing methods in external class libraries that contain parameters
             if (parameterTypeSymbol!.ConstructedFrom.SpecialType == SpecialType.System_Nullable_T &&
                 parameterTypeSymbol.TypeArguments.Length > 0)
             {

--- a/src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
+++ b/src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
@@ -1,7 +1,6 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Security.Policy;
 using Microsoft.CodeAnalysis.Testing;
 using VerifyCS = Microsoft.AspNetCore.Analyzers.Verifiers.CSharpAnalyzerVerifier<Microsoft.AspNetCore.Analyzers.RouteHandlers.RouteHandlerAnalyzer>;
 

